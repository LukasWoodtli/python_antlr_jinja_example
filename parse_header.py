from antlr4.error.ErrorListener import ErrorListener

from header_model.model import HeaderFileModel, ACCESS_SPECIFIER_PUBLIC, HeaderFileClassModel, HeaderFileMethodModel, \
    CppOnlyElement, HeaderFileVariableModel, ACCESS_SPECIFIER_PRIVATE, ACCESS_SPECIFIER_PROTECTED
from antlr4.BufferedTokenStream import BufferedTokenStream

from parsing.HeaderFileParser import *
from parsing.HeaderFileLexer import HeaderFileLexer
from parsing.HeaderFileListener import HeaderFileListener


class HeaderFileErrorListener(ErrorListener):

    def __init__(self) -> None:
        super(HeaderFileErrorListener, self).__init__()

    def syntaxError(self, recognizer, offending_symbol, line, column, msg, e):
        super().syntaxError(recognizer, offending_symbol, line, column, msg, e)
        raise Exception(f"Error at line {line}:{column} {msg}")


class HeaderFileModelCreator(HeaderFileListener):
    def __init__(self, parser):
        self.parser = parser
        self.header_file_model = HeaderFileModel()
        self.current_class = None
        self.current_access_modifier = ACCESS_SPECIFIER_PUBLIC

    def _get_source_location(self, ctx):
        return f"{self.header_file_model.name}: {ctx.start.line}, {ctx.start.column}"

# This helps debugging
#     def __getattribute__(self, name):
#         attr = object.__getattribute__(self, name)
#         if hasattr(attr, '__call__'):
#             def newfunc(*args, **kwargs):
#                 print('before calling %s with args %s and %s' % (attr.__name__, str(args), str(kwargs)))
#                 result = attr(*args, **kwargs)
#                 #print('done calling %s' % attr.__name__)
#                 return result
#
#             return newfunc
#         else:
#             return attr

    def get_model(self) -> HeaderFileModel:
        return self.header_file_model

    @staticmethod
    def _get_original_source_code(element):
        stream = element.start.getInputStream()
        arg_list_start = element.start.start
        arg_list_end = element.stop.stop
        element_source_code = stream.strdata[arg_list_start:arg_list_end + 1]
        return element_source_code

    def enterPreprocessorDirective(self, context: HeaderFileParser.PreprocessorDirectiveContext):
        self.header_file_model.add_statement(CppOnlyElement(context.getText().strip("\n")))

    def enterPreprocessorInclude(self, context: HeaderFileParser.PreprocessorIncludeContext):
        self.header_file_model.add_include(context.getText().strip("\n"))

    def enterClassDefinition(self, context: HeaderFileParser.ClassDefinitionContext):
        assert self.current_class == None, f"No nested classes allowed: {self._get_source_location(context)}"
        base_class = context.classHead().baseClassName()
        base_class_name = base_class.getText() if base_class else ""
        self.current_class = HeaderFileClassModel(context.classHead().className().getText(),
                                               base_class_name)

    def exitClassDefinition(self, context: HeaderFileParser.ClassDefinitionContext):
        self.header_file_model.add_class(self.current_class)

    def exitPublicAccessSpecifier(self, context):
        self.current_access_modifier = ACCESS_SPECIFIER_PUBLIC

    def exitProtectedAccessSpecifier(self, context):
        self.current_access_modifier = ACCESS_SPECIFIER_PROTECTED

    def exitPrivateAccessSpecifier(self, context):
        self.current_access_modifier = ACCESS_SPECIFIER_PRIVATE

    def exitMemberFunctionDefinition(self, context: HeaderFileParser.MemberFunctionDefinitionContext):
        assert self.current_class, f"No class defined: {self._get_source_location(context)}"

        is_const = True if context.constMethodQualifier() else False
        is_static = True if context.staticMethodQualifier() else False
        has_override = True if context.overrideMethodQualifier() else False
        return_type = self._get_original_source_code(context.returnType())
        function_name = context.functionName().getText()
        arg_list = context.functionArgumentList()

        arg_list_text = self._get_original_source_code(arg_list) if arg_list else ""

        method = HeaderFileMethodModel(return_type,
                                    function_name,
                                    self.current_access_modifier,
                                    is_const,
                                    is_static,
                                    has_override,
                                    arg_list_text)
        self.current_class.add_method(method)

    def exitMemberVariableDefinition(self, context: HeaderFileParser.MemberVariableDefinitionContext):
        assert self.current_class, f"No class defined: {self._get_source_location(context)}"

        access_modifier = self.current_access_modifier
        variable_context = context.variableDefinition()
        static = True if variable_context.staticVariableQualifier() else False
        const = True if variable_context.constVariableQualifier() else False
        type_name = variable_context.typeName().getText()
        variable_name = variable_context.variableName().getText()
        default_value = variable_context.defaultValueAssignment()
        default_value_assignment = self._get_original_source_code(default_value) if default_value else ""

        member_variable = HeaderFileVariableModel(access_modifier, static, const, type_name, variable_name, default_value_assignment)
        self.current_class.add_member_variable(member_variable)

    def exitTypedef(self, context: HeaderFileParser.TypedefContext):
        typedef = self._get_original_source_code(context)
        self.header_file_model.add_typedefs(typedef)

    def exitUsingDeclaration(self, context:HeaderFileParser.UsingDeclarationContext):
        using = self._get_original_source_code(context)
        self.header_file_model.add_using_declaration(using)

    def exitForwardDeclaration(self, ctx: HeaderFileParser.ForwardDeclarationContext):
        decl = self._get_original_source_code(ctx)
        self.header_file_model.add_forward_declaration(decl)

    def _add_type_for_forward_declaration(self, context: HeaderFileParser.ValueTypeNameContext):
        if self.current_class:
            type_name = ""
            for element in context.fullQualifiedTypeNameElements():
                type_name += self._get_original_source_code(element)

            self.current_class.add_forward_declaration(type_name)

    def exitFriendClassDeclaration(self, ctx:HeaderFileParser.FriendClassDeclarationContext):
        self.current_class.add_friend_decl(self._get_original_source_code(ctx))

    def exitPointerTypeName(self, context: HeaderFileParser.PointerTypeNameContext):
        self._add_type_for_forward_declaration(context.valueTypeName())

    def exitReferenceTypeName(self, context: HeaderFileParser.PointerTypeNameContext):
        self._add_type_for_forward_declaration(context.valueTypeName())


def parse_header_file(filename):
    print(f"Parsing: {filename}")
    input_stream = FileStream(filename, encoding='iso-8859-1')
    lexer = HeaderFileLexer(input_stream)
    lexer.removeErrorListeners()
    lexer.addErrorListener(HeaderFileErrorListener())
    stream = BufferedTokenStream.BufferedTokenStream(lexer)
    parser = HeaderFileParser(stream)
    parser.removeErrorListeners()
    parser.addErrorListener(HeaderFileErrorListener())

    try:
        tree = parser.headerFile()
    except Exception as err:
        error_msg = f"Error occured while parsing: {filename}\n\t"
        raise Exception(error_msg + str(err))

    listener = HeaderFileModelCreator(parser)
    walker = ParseTreeWalker()
    walker.walk(listener, tree)

    return listener.get_model()
