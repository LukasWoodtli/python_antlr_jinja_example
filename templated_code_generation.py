import os
import tempfile

from jinja2 import Environment, FileSystemLoader

from header_model.model import ACCESS_SPECIFIER_PROTECTED, ACCESS_SPECIFIER_PUBLIC, ACCESS_SPECIFIER_PRIVATE


class CppGenerationHelper:
    def __init__(self, model):
        self.model = model
        self.TYPE_INCLUDE_MAP = {
            "flags": '#include "flags/flags.h"',
        }

    def name(self):
        return self.model.get_class().get_name()

    def gen_class_declaration(self):
        decl = f"class {self.name()}"
        base_class = self.model.get_class().get_base_class()
        if base_class:
            decl += f" : public {base_class}"
        return decl

    def gen_system_includes_from_header(self):
        includes = [inc for inc in self.model.get_includes() if "<" in inc]
        if "#include <string>" not in includes:
            includes.insert(0, "#include <string>")
        return includes

    def _map_forward_declarations_to_includes(self):
        tokens = self.model.get_class().get_forward_declarations()

        includes = self._map_types_to_includes(tokens)
        return includes

    def _map_return_types_to_includes(self):
        methods = self.model.get_class().get_methods()
        tokens = [method.return_type for method in methods]

        includes = self._map_types_to_includes(tokens)

        return includes

    def _map_types_to_includes(self, tokens):
        includes = []
        for token in tokens:
            if token in self.TYPE_INCLUDE_MAP:
                includes.append(self.TYPE_INCLUDE_MAP[token])
        return includes

    def _map_argument_type_to_includes(self):
        all_includes = []
        methods = self.model.get_class().get_methods()
        for method in methods:
            tokens = method.argument_list.split()

            includes = self._map_types_to_includes(tokens)
            all_includes.extend(includes)

        return all_includes

    def _gen_project_includes_from_header(self):
        includes = [inc for inc in self.model.get_includes() if '"' in inc]

        includes.extend(self._map_forward_declarations_to_includes())
        includes.extend(self._map_return_types_to_includes())
        includes.extend(self._map_argument_type_to_includes())
        base_class = self.model.get_class().get_base_class()
        if base_class:
            includes.append(f'#include "{base_class}.h"')
        return includes

    def gen_includes(self):
        includes = self._gen_project_includes_from_header()
        return includes

    def _gen_forward_declarations_from_argument_types(self):
        decls = self.model.get_class().get_forward_declarations()
        ret = []
        for declaration in decls:
            ret.append(f"class {declaration};")
        return ret

    @staticmethod
    def _remove_duplicates(forward_decls):
        return list(dict.fromkeys(forward_decls))

    def gen_forward_declarations(self):
        forward_decls = self._gen_forward_declarations_from_argument_types()
        forward_decls.extend(self.model.get_forward_declaration())
        forward_decls = self._remove_duplicates(forward_decls)
        return forward_decls

    def gen_typedefs(self):
        return "\n".join(self.model.get_typedefs())

    def gen_friend_declarations(self):
        friend_decls = self.model.get_class().get_friend_decls()
        return "\n".join(friend_decls)

    @staticmethod
    def _map_types(argument_list):
        ARGUMENT_TYPE_MAP = {"long": "long int"}
        for k, v in ARGUMENT_TYPE_MAP.items():
            argument_list = argument_list.replace(k, v)
        return argument_list

    def gen_get_class_name(self):
        override = "override" if self.model.get_class().get_base_class() else ""
        code = "virtual std::string className() const " + override + ' {return "' + self.name() + '";}'
        return code

    def gen_cpp_methods(self):
        klass = self.model.get_class()
        visibility = "PUBLIC"
        for method in klass.get_methods():
            const_txt = " const " if method.is_const else ""
            override_txt = ""
            if method.name == "init":
                if len(method.argument_list) > 0 and not method.has_override:
                    override_txt = ""
                elif klass.get_base_class():
                    override_txt = "override"
            elif method.has_override:  # if in source: force 'override'
                override_txt = "override"

            static_or_virtual_txt = "static " if method.is_static else "virtual"
            return_type = self._map_types(method.return_type)
            visibility_code = ""
            if method.access_specifier != visibility:
                visibility_code = method.access_specifier.lower() + ":\n"
                visibility = method.access_specifier
            method_code = f"{visibility_code}\t{static_or_virtual_txt} {return_type} {method.name}({self._map_types(method.argument_list)}){const_txt} {override_txt};\n"
            yield method_code

    def has_init_method(self):
        return self.model.get_class().has_init_method()

    def _gen_member_variables(self, access_specifier):
        variables = self.model.get_class().get_member_variables()
        variables = [var for var in variables if var.access == access_specifier]
        code = ""
        for var in variables:
            static = "static" if var.static else ""
            const = "const" if var.const else ""
            type_name = self._map_types(var.type_name)
            code += f'{static} {const} {type_name} {var.variable_name} {var.default_value_assignment};'
        return code

    def gen_public_member_variables(self):
        return self._gen_member_variables(ACCESS_SPECIFIER_PUBLIC)

    def gen_protected_member_variables(self):
        return self._gen_member_variables(ACCESS_SPECIFIER_PROTECTED)

    def gen_private_member_variables(self):
        return self._gen_member_variables(ACCESS_SPECIFIER_PRIVATE)

    def gen_using_declarations(self):
        using_decls = self.model.get_using_declarations()
        return "\n".join(using_decls)


DIR_PATH = os.path.dirname(os.path.realpath(__file__))
TEMPLATES_FOLDER = os.path.join(DIR_PATH, "templates")
MODULE_DIR = os.path.join(tempfile.gettempdir(), os.path.split(DIR_PATH)[-1])
ENCODING = 'ascii'
env = Environment(
    line_statement_prefix='%',
    line_comment_prefix='##',
    loader=FileSystemLoader(TEMPLATES_FOLDER, encoding=ENCODING),
    autoescape=False,
    auto_reload=True,  # set to false for performance improvement in production
    keep_trailing_newline=True,
    trim_blocks=False,
    lstrip_blocks=False
)


def generate_h_file(model):
    cpp_model = CppGenerationHelper(model)
    return _generate_output_file("CppHeaderTemplate.jinja2", cpp_model)


def _generate_output_file(template_file, cpp_model):
    template = env.get_template(template_file)
    templated_out_f = template.render(cpp_model=cpp_model)
    return templated_out_f
