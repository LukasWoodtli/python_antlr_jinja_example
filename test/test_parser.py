import os

from approvaltests.approvals import verify

import templated_code_generation
from parse_header import parse_header_file


input_file = os.path.dirname(__file__)
input_file = os.path.join(input_file, "input_files", "TestInputFile.h")


def test_parse_file():
    model = parse_header_file(input_file)
    assert model.get_class().get_name() == "TestExampleClass"


def test_generate_files():
    model = parse_header_file(input_file)
    generated_code = templated_code_generation.generate_h_file(model)

    verify(generated_code)

