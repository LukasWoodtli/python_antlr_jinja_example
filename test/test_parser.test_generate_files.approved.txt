#pragma once


#include <string>
#include <chrono>
#include "TestExampleClass.h"
#include "MyOtherTestingClass.h"
#include "flags/flags.h"
#include "flags/flags.h"
#include "BaseTestClass.h"
using MyFancyFn = std::function<bool(int*)>;
class date;
class array;
class AmisDate;
class abc::def::foo;
class flags;
class Hello;
class Hallo;
class World;
class MyForwardDeclaration;
typedef std::pair<double, int> MyPair;


class TestExampleClass : public BaseTestClass
{
public:
    friend class MyGoodFriend;
	virtual ~TestExampleClass();

	static SpTestExampleClass CreateTestExampleClass();

	virtual std::string className() const override {return "TestExampleClass";}

	void init();
       Hello* hello ;  double someBool ;  BigThing reallyBig ;

    

protected:
    

    TestExampleClass();

public:
	virtual int createArray(
		const date& theDate,
	                                    const date& otherDate,
	                                    long int hello,
	                                    array* theArray) ;

	virtual Nowrapper int getArray(
		const date* theDate,
	                                  const AmisDate* otherDate,
	                                  array* theArray,
	                                  long int type = 0,
	                                  date* oneMoreDate = nullptr,
	                                  bool aBoolean= false) ;

	virtual int getSomething(
		array* theArray) ;

	virtual double getSomethingElse(
		int i = C42) ;

	virtual const abc::def::foo& getFancyThing() const  ;

	static  long int getMoreStuff(
		bool theBool) ;

	virtual flags flags() const  ;

	virtual void setFlags(
		const flags& f) ;

	virtual bool isFSomethingTrue() ;

};

